### PR TITLE
Process GRO-ed SKBs in a correct way.

### PR DESCRIPTION
**Pull request by keshonok**
This fixes kernel crashes in multiple places as noted in issues
#430, #431, #432, #433, #435, #437.

The core issue was the ss_skb_split() is not supposed to work with
GRO-ed SKBs that has data in skb_shinfo(skb)->frag_list. Splitting
of those SKBs resulted in the appearance of all kinds of invalid
SKBs that caused total havoc in the kernel. It also triggered #436
that will be fixed separately.

### CODE REVIEW

File: tempesta_fw/sock.c [hunk 1]
```diff
@@ -210,13 +210,13 @@ ss_send(struct sock *sk, SsSkbList *skb_list, bool pass_skb)
 		ss_skb_queue_head_init(&sw.skb_list);
 		for (skb = ss_skb_peek(skb_list); skb; skb = ss_skb_next(skb)) {
 			/* tcp_transmit_skb() will clone the skb. */
-			skb_copy = pskb_copy_for_clone(skb, GFP_ATOMIC);
-			if (!skb_copy) {
+			twin_skb = pskb_copy_for_clone(skb, GFP_ATOMIC);
```
Comment (keshonok) [inline tempesta_fw/sock.c]: I renamed it as I experimented with different SKB copy functions, one of which is `skb_copy()`. That caused the name conflict. I suggest that we keep this change.

File: tempesta_fw/sock.c [hunk 2]
```diff
@@ -482,15 +485,39 @@ ss_tcp_process_data(struct sock *sk)
 		skb_dst_drop(skb);
 		skb->dev = NULL;
```
Comment (keshonok) [inline tempesta_fw/sock.c]: It appears that `skb->dev = NULL` is always done by the Linux stack in `tcp_v4_do_rcv()`: http://lxr.free-electrons.com/source/net/ipv4/tcp_ipv4.c?v=4.1#L1626

File: tempesta_fw/sock.c [hunk 3]
```diff
@@ -482,15 +485,39 @@ ss_tcp_process_data(struct sock *sk)
 		skb_dst_drop(skb);
```
Comment (keshonok) [inline tempesta_fw/sock.c]: Shouldn't this be done in the egress path, i.e. in `ss_do_send()`? Until that time it doesn't matter at all as the SKB is taken out of the Linux TCP/IP stack and orphaned.

Comment (krizhanovsky) [inline tempesta_fw/sock.c]: Maybe so, I don't mind to move it to ```ss_do_send()```. The intention to add the call here was to "forget" routing information for ingress skb and process it further the same way as locally generated skbs.

File: tempesta_fw/sock.c (hunk 2 above)
Comment (krizhanovsky) [inline tempesta_fw/sock.c]: Yeah, it seems it's safe to remove the line.

File: tempesta_fw/sock.c [hunk 4]
```diff
@@ -482,15 +485,39 @@ ss_tcp_process_data(struct sock *sk)
...
+			if (TCP_SKB_CB(skb)->tcp_flags & TCPHDR_FIN) {
+				TCP_SKB_CB(skb)->tcp_flags &= ~TCPHDR_FIN;
+				TCP_SKB_CB(last_skb)->tcp_flags |= TCPHDR_FIN;
+			}
+			skb_shinfo(skb)->frag_list = NULL;
+		}
+
 		off = tp->copied_seq - TCP_SKB_CB(skb)->seq;
-		if (tcp_hdr(skb)->syn)
+		if (TCP_SKB_CB(skb)->tcp_flags & TCPHDR_SYN)
```
Comment (krizhanovsky) [inline tempesta_fw/sock.c]: Why the change is required?

File: tempesta_fw/sock.c [hunk 5]
```diff
@@ -482,15 +485,39 @@ ss_tcp_process_data(struct sock *sk)
...
 			off--;
+repeat:
+		/* SKB may be freed in processing. Save the flag. */
+		tcp_fin = TCP_SKB_CB(skb)->tcp_flags & TCPHDR_FIN;
```
Comment (krizhanovsky) [inline tempesta_fw/sock.c]: Why don't you use ```tcp_hdr(skb)->fin``` ?

File: tempesta_fw/sock.c [hunk 6]
```diff
@@ -482,15 +485,39 @@ ss_tcp_process_data(struct sock *sk)
...
+		 * SKBs for processing. Move the FIN flag if necessary.
+		 */
+		frag_list = skb_shinfo(skb)->frag_list;
+		if (frag_list) {
+			struct sk_buff *frag_skb, *last_skb = frag_list;
+			skb_walk_frags(skb, frag_skb) {
+				frag_skb->nohdr = 0;
+				ss_skb_adjust_data_len(skb, -frag_skb->len);
+				last_skb = frag_skb;
+			}
```
Comment (krizhanovsky) [inline tempesta_fw/sock.c]: Then ```ss_skb_process()``` shall not walk through the fragments. And this is good to do in this patch to clearly express the logic movement.

File: tempesta_fw/sock.c [hunk 7]
```diff
@@ -542,13 +542,6 @@ ss_tcp_process_data(struct sock *sk)
...
 			processed += count;
 
-			if (frag_list) {
-				off = 0;
-				skb = frag_list;
-				frag_list = frag_list->next;
-				skb->next = skb->prev = NULL;
-				goto repeat;
-			}
 			if (tcp_fin) {
```
Comment (krizhanovsky) [inline tempesta_fw/sock.c]: So if FIN is actually set in the last fragment we don't process all the fragments but first.

File: tempesta_fw/ss_skb.c [hunk 8]
```diff
@@ -164,58 +162,65 @@ __lookup_pgfrag_room(struct sk_buff *skb, int len)
...
  * Somewhat like skb_shift().
+ * Make room for @n fragments starting with slot @from.
  *
- * Beware: @from can be equal to MAX_SKB_FRAGS if we need to insert a new
- * fragment after the last one.
+ * Beware: @from can be equal to MAX_SKB_FRAGS when a new fragment
+ * is inserted after the last one.
  */
 static int
 __extend_pgfrags(struct sk_buff *skb, struct sk_buff *pskb, int from, int n)
```
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: pskb is unused argument

File: tempesta_fw/ss_skb.c [hunk 9]
```diff
@@ -527,70 +575,102 @@ static int
...
 
+/*
+ * Move an SKB from frag_list to the proper place in the list of SKB
+ * for an HTTP message.
+ *
+ * During the manipulations on SKB fragments a new SKB may be created
+ * to hold new or relocated fragments. The new SKB is stored in the
+ * original SKB's frag_list. After the manipulations on fragments are
+ * completed, the new SKB is inserted after the original SKB in the
+ * list of SKBs for an HTTP message.
```
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: The comment is unclear. What is "proper place"? Currently there are strict order for skb processing, so fragments were also at proper place in sense of processing order. It's clear from the code that skb is moved from fragment to Tempesta's list of skbs. However, it's absolutely unclear why fragments don't work for us anymore and we need to linearize the skb list.

File: tempesta_fw/ss_skb.c [hunk 10]
```diff
@@ -237,6 +246,7 @@ __new_pgfrag(struct sk_buff *skb, struct sk_buff *pskb, int size, int i,
 			return -ENOMEM;
 	}
 
+	/* Make room for @shift fragments starting with slot @i. */
 	if (__extend_pgfrags(skb, pskb, i, shift)) {
```
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: pskb is actually unused, but still passed as an argument.

File: tempesta_fw/ss_skb.c [hunk 11]
```diff
@@ -299,36 +316,43 @@ __split_linear_data(struct sk_buff *skb, struct sk_buff *pskb,
...
 			return NULL;
 	} else {
-		/*
-		 * Probably we should delete the whole second part
-		 * of the linear data or move it to page fragment.
-		 */
-		tail_len += len;
-		BUG_ON(tail_len < 0);
-		if (tail_len && __extend_pgfrags(skb, pskb, 0, 1))
+		if (__extend_pgfrags(skb, pskb, 0, 1))
```
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: Why do you extend fragments array if we don't need to insert a new fragment? If ```tail_len == 0```, i.e. we delete the whole tail, then this is just straightforward head length adjustment.

File: tempesta_fw/ss_skb.c [hunk 12]
```diff
@@ -345,96 +369,106 @@ __split_linear_data(struct sk_buff *skb, struct sk_buff *pskb,
...
- * @return pointer to the new data room or NULL on failure.
+ * Note: @off is always within the borders of fragment @i. It can
+ * point at the start of a fragment, but it can never point at the
+ * location right after the end of a fragment. In other words, @off
+ * can be zero, but it can not be equal to the size of fragment @i.
+ *
+ * @return pointer to the room for new data, or NULL on failure.
  */
 static void *
 __split_pgfrag_add(struct sk_buff *skb, struct sk_buff *pskb, int i, int off,
```
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: Unused ```pskb``` again

File: tempesta_fw/ss_skb.c [hunk 13]
```diff
@@ -345,96 +369,106 @@ __split_linear_data(struct sk_buff *skb, struct sk_buff *pskb,
...
 {
-	int split, dst_i = i + !!off;
-	skb_frag_t *frag_dst, *frag = &skb_shinfo(skb)->frags[i];
+	int tail_len;
 	struct sk_buff *skb_dst;
+	skb_frag_t *frag_dst, *frag = &skb_shinfo(skb)->frags[i];
 
+	SS_DBG("%s: skb [%p] i [%d] off [%d] len [%d] fragsize [%d]\n",
+		__func__, skb, i, off, len, skb_frag_size(frag));
 	BUG_ON((off < 0) || (off >= skb_frag_size(frag)));
```
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: Please add ```BUG_ON(off + len < skb_frag_size(frag))``` to be sure that we never call the function with ```off``` pointing to the end of fragment.

File: tempesta_fw/ss_skb.c [hunk 14]
```diff
@@ -345,96 +369,106 @@ __split_linear_data(struct sk_buff *skb, struct sk_buff *pskb,
...
 
 	/*
-	 * Try to append data to a page fragment. If 'off' is zero,
-	 * then try to append data to a preceding page fragment
+	 * Try to append data to the page fragment. If @off is zero,
+	 * then try to append data to the preceding page fragment
 	 * if there's any. Go for other solutions if there's no room.
 	 */
-	if (!split && (off || i)) {
+	if (off || i) {
```
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: The condition is wrong: we're entering to the branch ever if we must insert data into middle of the fragment. Meantime, if shall be ```if (!off && i)```, i.e. try to append the data to previous fragment (i != 0) if we need to add data just before the fragment (!off).

Since now we don't handle case adding data just after i'th fragment, the internals of the branch must also be amended - ```frag_dst``` is always ```frag```, so we actually just should check current ```frag``` and can do this in the condition above.

File: tempesta_fw/ss_skb.c [hunk 15]
```diff
@@ -345,96 +369,106 @@ __split_linear_data(struct sk_buff *skb, struct sk_buff *pskb,
...
 }
 
 /**
- * Delete @len (the value is postive now) bytes from @frag.
- * @return pointer just after deleted fragment.
+ * Delete @len (the value is positive now) bytes from @frag.
+ * @return a pointer at the location right after the deleted fragment.
  */
 static void *
 __split_pgfrag_del(struct sk_buff *skb, struct sk_buff *pskb, int i, int off,
```
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: unused ```pskb```

File: tempesta_fw/ss_skb.c [hunk 16]
```diff
@@ -461,36 +500,49 @@ __split_pgfrag_del(struct sk_buff *skb, struct sk_buff *pskb, int i, int off,
...
 		ss_skb_adjust_data_len(skb, -len);
 		return skb_frag_address(frag);
 	}
+	/* Quick path: delete the tail part of a fragment. */
+	if (off + len == skb_frag_size(frag)) {
+		skb_frag_size_sub(frag, len);
+		ss_skb_adjust_data_len(skb, -len);
+		++i;
+		goto lookup_next_ptr;
+	}
```
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: Good catch!

File: tempesta_fw/ss_skb.c [hunk 17]
```diff
@@ -498,16 +550,12 @@ __split_pgfrag_del(struct sk_buff *skb, struct sk_buff *pskb, int i, int off,
...
-	/* Try to find next data chunk after deleted fragment. */
+	/* Get the next fragment after the deleted fragment. */
 	if (i < si->nr_frags)
 		return skb_frag_address(&si->frags[i]);
-	skb_dst = skb_shinfo(pskb ? : skb)->frag_list;
-	if (skb_headlen(skb_dst))
-		return skb_dst->data;
-	if (skb_shinfo(skb_dst)->nr_frags)
-		return skb_frag_address(&skb_shinfo(skb_dst)->frags[0]);
-	/* ...not found, return invalid address try next skb. */
```
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: Hmmm, it seems you abandoned skb fragments support just to remove the piece of code.... I don't see problems with the code while really ugly ```return (void *)1``` is still there.

File: tempesta_fw/ss_skb.c [hunk 18]
```diff
@@ -527,70 +575,102 @@ static int
...
+			it->ptr = __split_pgfrag(skb, pskb, i, offset, len);
 			goto done;
 		}
 	}
 
-	skb_walk_frags(skb, frag_i) {
-		int r = __skb_fragment(frag_i, skb, pspt, len, it);
-		if (r != -ENOENT)
-			return r;
-	}
```
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: Why do you delete the code? ```ss_skb_defraglist()``` actually doesn't move the whole fragments list to ```TFW_SKB_CB()``` list. If you'd move the list, then you need to convert skb fragment list to Tempesta's list, which use different skb fields to organize the lists. This is double list traversals (firstly to convert the list and secondly to process the list) and there is no sense for the overhead, whereas the lists are typically short.

File: tempesta_fw/ss_skb.c [hunk 19]
```diff
@@ -601,19 +681,18 @@ __skb_fragment(struct sk_buff *skb, struct sk_buff *pskb, char *pspt,
...
-ss_skb_get_room(struct sk_buff *skb, char *pspt, unsigned int len, TfwStr *it)
+ss_skb_get_room(SsSkbList *head, struct sk_buff *skb,
+		char *pspt, unsigned int len, TfwStr *it)
 {
-	if (unlikely(len > PAGE_SIZE)) {
-		SS_WARN("Trying to get too large skb room, size=%u\n", len);
-		return -E2BIG;
-	}
+	int r = __skb_fragment(skb, NULL, pspt, len, it);
+	ss_skb_defraglist(head, skb);
```
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: It's unclear why do you do this, but anyway there is no sense to put the new fragment for ```frag_list``` and move to the skb list. It's rather better to pass ```head``` to ```__skb_fragment()```.

File: tempesta_fw/ss_skb.c [hunk 20]
```diff
@@ -498,16 +550,12 @@ __split_pgfrag_del(struct sk_buff *skb, struct sk_buff *pskb, int i, int off,
...
+	/* Get the next fragment after the deleted fragment. */
 	if (i < si->nr_frags)
 		return skb_frag_address(&si->frags[i]);
-	skb_dst = skb_shinfo(pskb ? : skb)->frag_list;
-	if (skb_headlen(skb_dst))
-		return skb_dst->data;
-	if (skb_shinfo(skb_dst)->nr_frags)
-		return skb_frag_address(&skb_shinfo(skb_dst)->frags[0]);
-	/* ...not found, return invalid address try next skb. */
+	/* Return invalid address, try next skb. */
```
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: We spent a lot of time to understand were do we need the return value, so it'd be good to reference ```ss_skb_cutoff_data()``` in the comment.

File: tempesta_fw/ss_skb.c [hunk 21]
```diff
@@ -527,70 +575,102 @@ static int
...
+	skb_frag_list_init(skb);
+}
+
+static inline void *
+ss_skb_data_address(struct sk_buff *skb)
+{
+	if (skb_headlen(skb))
+		return skb->data;
+	if (skb_shinfo(skb)->nr_frags)
+		return skb_frag_address(&skb_shinfo(skb)->frags[0]);
```
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: I'm not sure is it possible that skb has only ```frag_list```, but the function should either lookup the data in ```skb->frag_list``` or make an assertion that this is impossible case.

File: tempesta_fw/sock.c [hunk 22]
```diff
@@ -509,38 +533,59 @@ ss_tcp_process_data(struct sock *sk)
...
+				}
+				SS_DBG("Stop processing: sk %p\n", sk);
+				break;
+			}
+			if (frag_list) {
+				off = 0;
+				skb = frag_list;
+				frag_list = frag_list->next;
+				skb->next = NULL;
+				goto repeat;
```
Comment (krizhanovsky) [inline tempesta_fw/sock.c]: This is internal loop on skb fragments and you have 3 places where you do iterations over the list. The function also becomes to long, so it must be reworked: current loop over ```sk->sk_receive_queue``` stays in the function while the loop over ```skb->frag_list``` must be moved to other function and replaced by normal loop construction instead of goto. Probably the body of the internal loop is good to move to 3rd function.

File: tempesta_fw/ss_skb.c (hunk 11 above)
Comment (keshonok) [inline tempesta_fw/ss_skb.c]: Here is the case where we know for sure that we need to insert a new fragment. There's a case where a whole tail is removed from the linear part. That case is processed before this one, not shown in this patch.

Also, `tail_len` can never be zero. If it was that would mean that we've got `pspt` pointing at the end of data in the linear area, because `tail_len` is calculated as `tail_len = (char *)skb_tail_pointer(skb) - pspt;`

Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: Yeah, you're right

File: tempesta_fw/ss_skb.c (hunk 19 above)
Comment (keshonok) [inline tempesta_fw/ss_skb.c]: In the manipulations on the SKB it may be necessary to create a new SKB to hold extra fragments. The new SKB is created in `__extend_pgfrags()`. The new SKB has to be inserted into the SKB list for an HTTP message in correct place, after the current SKB that is being processed. That operation requires the list head. So the list head will have to be passed through the whole call chain in `ss_skb.c` all the way down to `__extend_pgfrags()`. 

Is that what we want to do?

Instead, I temporarily store it in SKB's `frag_list` as now we don't have data there. When the control is returned back to one of the top functions in the call chain that has the pointer to the list head, the SKB is moved to the SKB list for the HTTP message.

File: tempesta_fw/ss_skb.c (hunk 21 above)
Comment (keshonok) [inline tempesta_fw/ss_skb.c]: In this sequence of patches we don't have data in `frag_list` anymore, guuaranteed. I will put a `BUG_ON()` to assert that.

File: tempesta_fw/ss_skb.c (hunk 9 above)
Comment (keshonok) [inline tempesta_fw/ss_skb.c]: In this sequence of patches we don't have data in `frag_list` of an SKB anymore. `frag_list` is used as the palce to temporarily store the new SKB that is created in the process of manipulating the SKB's fragments and data in those fragments. The "proper" place is after the SKB that just had been processed. That is explained in the lengthy comment above.

Please, see also the comment below: https://github.com/natsys/tempesta/pull/440/files#r58870927

File: tempesta_fw/ss_skb.c (hunk 18 above)
Comment (keshonok) [inline tempesta_fw/ss_skb.c]: In this series of patches we don't have data in `frag_list` anymore.
`ss_skb_defraglist()` moves just the new SKB that was created while doing an add or delete operation on data in the SKB. There can be only one such SKB. There's no double list traversal.
Please see also the comments below on the same topic.

File: tempesta_fw/ss_skb.c (hunk 17 above)
Comment (keshonok) [inline tempesta_fw/ss_skb.c]: We get here when we don't have knowledge of where the next data is. Thus this `return (void *)1`. How do you propose we can eliminate that?

Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: I don't propose to remove the code. Rather I argue to not to remove ```frag_list`` since it has no sense and do better commenting for the code.

File: tempesta_fw/ss_skb.c (hunk 18 above)
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: Ok. But don't forget assertions that there is no frag_list

File: tempesta_fw/ss_skb.c (hunk 9 above)
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: That's not the right technique to pass function return value through list of fragments. The function should be removed and the new fragment must be handled properly.

File: tempesta_fw/ss_skb.c (hunk 21 above)
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: Yeah, BUG_ON is very wished

File: tempesta_fw/ss_skb.c (hunk 19 above)
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: I don't see a problem to pass list head by a function argument. Also you can use some return value.

File: tempesta_fw/ss_skb.c [hunk 23]
```diff
@@ -186,8 +186,9 @@ __extend_pgfrags(struct sk_buff *skb, struct sk_buff *pskb, int from, int n)
 		 * Allocate a new SKB to hold @n_excess fragments.
 		 * Put it on frag_list of the current SKB.
 		 */
-		if ((nskb = alloc_skb(0, GFP_ATOMIC)) == NULL)
+		if ((nskb = alloc_skb(MAX_TCP_HEADER, GFP_ATOMIC)) == NULL)
```
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: Why do we need TCP header in the fragment?

File: tempesta_fw/ss_skb.c [hunk 24]
```diff
@@ -164,129 +162,155 @@ __lookup_pgfrag_room(struct sk_buff *skb, int len)
...
-			psi->frag_list = skb_frag;
-		}
+		/*
+		 * Allocate a new SKB to hold @n_excess fragments.
+		 * Put it on frag_list of the current SKB.
+		 */
+		nskb = alloc_skb(MAX_TCP_HEADER, GFP_ATOMIC);
+		if (nskb == NULL)
+			return -ENOMEM;
+		skb_reserve(nskb, MAX_TCP_HEADER);
```
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: Why do we need to be able to place TCP header in the fragment?

File: tempesta_fw/ss_skb.c [hunk 25]
```diff
@@ -207,9 +208,10 @@ __extend_pgfrags(struct sk_buff *skb, struct sk_buff *pskb, int from, int n)
 		}
 		skb_shinfo(nskb)->nr_frags += n_excess;
 	}
+
 	/* Make room for @n fragments in the SKB. */
 	n_shift = si->nr_frags - from - n_excess;
-	if (n_shift)
+	if (n_shift > 0)
```
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: Probably it has sense to add BUG_ON(n_shift < 0)

File: tempesta_fw/ss_skb.c [hunk 26]
```diff
@@ -164,129 +162,155 @@ __lookup_pgfrag_room(struct sk_buff *skb, int len)
...
+		skb_shinfo(nskb)->nr_frags += n_excess;
+		it->skb = nskb;
 	}
 
-	memmove(&si->frags[from + n], &si->frags[from],
-		(si->nr_frags - from - n_frag) * sizeof(skb_frag_t));
-	si->nr_frags += n - n_frag;
+	/* Make room for @n fragments in the SKB. */
+	n_shift = si->nr_frags - from - n_excess;
+	if (n_shift > 0)
```
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: Probably it has sense to add BUG_ON(n_shift < 0)

File: tempesta_fw/ss_skb.c [hunk 27]
```diff
@@ -461,136 +512,185 @@ __split_pgfrag_del(struct sk_buff *skb, struct sk_buff *pskb, int i, int off,
...
+	/*
+	 * Use @it to hold the return values from __split_pgfrag()
+	 * and __split_linear_data(). @it->ptr, @it->skb, and
+	 * @it->flags may be set to actual values. If a new SKB is
+	 * allocated, then it is stored in @it->skb. @it->ptr holds
+	 * the pointer either to data after the deleted data, or to
+	 * the area for new data. @it->flags indicates that @it->ptr
+	 * points to data in @it->skb.
+	 */
+	memset(it, 0, sizeof(TfwStr));
```
Comment (krizhanovsky) [inline tempesta_fw/ss_skb.c]: Why do you need this? If something is wrong, then we return error code. If not, then ```it```'s members aren't used.

Comment (keshonok) [inline tempesta_fw/ss_skb.c]: If the result of modification is positive (there was no error), that doesn't mean we have the relevant values in `it->skb` and `it->ptr`. `it->skb` is not set if the modification didn't require a new SKB. `it->ptr` is not set if it should be set to an area in the next SKB that is not known at that time. So, in that case, we need to see that whether the values have changed or not. For that reason we can't have garbage in `it` and need to make sure it's cleared before use.

The block of code below at the label `done` deals with interpreting the  values of `it->skb`, `it->ptr`, and it->flags`.

File: tempesta_fw/ss_skb.c (hunk 24 above)
Comment (keshonok) [inline tempesta_fw/ss_skb.c]: It turns out that when these SKBs get into TCP/IP stack for sending, the stack expects that SKBs have sufficient room to do an `skb_pull()`. If there's no headroom, it crashes in skb_pull().

The total absence of headroom is okay for SKB fragments (from `frag_list`), but it is not in this case.

File: tempesta_fw/sock.c [hunk 28]
```diff
@@ -432,6 +432,102 @@ ss_close_sync(struct sock *sk)
...
+			ss_skb_adjust_data_len(skb, -f_skb->len);
+		}
+		skb_shinfo(skb)->frag_list = NULL;
+	}
+	skb->next = frag_list;
+	frag_list = skb;
+
+	while (frag_list) {
+		skb = frag_list;
+		frag_list = frag_list->next;
```
Comment (krizhanovsky) [inline tempesta_fw/sock.c]: ```skb``` is never NULL, you assign ```skb``` to ```frag_list```, which is just a loop iterator, check it for NULL and reassign ```frag_list``` to ```skb``` - looks tricky and ugly. Why not to use plain for() loop?

**Comment by krizhanovsky** [comment]
Good to merge without https://github.com/natsys/tempesta/pull/440/commits/607ddc750907d753c325379010cb1679b0d66873
